### ISSUE TITLE
SoapClient::__setLocation support

### ISSUE THREAD

**Issue by Greg0** [phpro/soap-client#367]
Hi,
First of all thank you for your incredible library.

I have question for [SoapClient::__setLocation](https://www.php.net/manual/en/soapclient.setlocation.php) method.
How can I setup custom web service url location instead of using that returned in WSDL.
Should I use some middleware or something other?

**Comment by veewee** [comment]
The suggested way for changing the target URL is by using a HTTP plugin to overwrite the base URL:
http://docs.php-http.org/en/latest/plugins/request-uri-manipulations.html#request-uri-manipulations

Alternatively, you could unwind some of the factories and set the `__setLocation` on the client that is being used internally.

**Comment by veewee** [comment]
The question has been answered. Closing this issue.
Feel free to reopen or create a new one if your issue is not solved.

**Comment by tgerakitis** [comment]
@veewee I wanted to use this package for my latest implementation and I'm also stuck on this particular topic.

Setting the location is pretty common due to staging and different environments.
Probably my fault, but I don't understand your answer and how to implement setting the location using the link you provided..

Could you please explain a bit simpler?

**Comment by Greg0** [comment]
@tgerakitis Finally I achieved that by passing custom location as option

```php
$options = ['location' => 'http://your_custom_domain.com'];
$engine = ExtSoapEngineFactory::fromOptionsWithHandler(
            ExtSoapOptions::defaults($wsdlLocation, $options)
                ->withClassMap(ServiceClassmap::getCollection()),
            $handler
        );
```

**Comment by tgerakitis** [comment]
thank you @Greg0! :)

I tried templating the location in my `wsdl` file directly, 
your way is a much better approach because I don't need to manually edit the `wsdl` before using it.

**Comment by Greg0** [comment]
> I tried templating the location in my wsdl file directly,
your way is a much better approach because I don't need to manually edit the wsdl before using it.

Especially that not always we are the owner of WSDL file, so not always we can edit it :) 

All options passed like that directly goes to built-in sopa library, to `\SoapClient` construct:
https://www.php.net/manual/en/soapclient.construct.php

**Comment by veewee** [comment]
Ah yes, that is a good solution as well! Thanks for sharing.

> I don't really understand why this proposal idea was closed... It seems simple and non breaking to add an `$additionalOptions` parameter to the factory call like so.

See:
https://github.com/phpro/soap-client/blob/master/docs/cli/generate-clientfactory.md

You might not need "options" for your implementation.
We decided to generate a basic factory from which you can start configuring your exact needs. If you pass additional options, that would be perfectly fine. But you might also want to pass http or wsdls middlewares. Where do you draw the line of what should be configurable from the upstream code or from inside the factory?


We configure the factory from our container and even drop the wsdls part of the function to be perfectly open... 😉
